flair/embeddings.py

```diff
@@ -994,6 +1193,149 @@ def __str__(self):
+                    tokens_tensor = tokens_tensor.to(flair.device)
+
+                    hidden_states = self.model(tokens_tensor)[-1]
+
+                    subtoken_embeddings = _extract_embeddings(
+                        hidden_states=hidden_states,
+                        layers=self.layers,
+                        pooling_operation=self.pooling_operation,
+                    )
+
+                    final_subtoken_embedding = torch.cat(subtoken_embeddings)
+                    token.set_embedding(self.name, final_subtoken_embedding)
+
+
+class RoBERTaEmbeddings(TokenEmbeddings):
+    def __init__(
+        self,
+        model: str = "roberta.large",
+        layers: str = "-1",
+        pooling_operation: str = "first",
+    ):
+        """RoBERTa, as proposed by Liu et al. 2019.
+        :param model: name of RoBERTa model
+        :param layers: comma-separated list of layers
+        :param pooling_operation: defines pooling operation for subwords
+        """
+        super().__init__()
+
+        try:
+            self.model = torch.hub.load("pytorch/fairseq", model)
+        except:
+            log_line(log)
+            log.warning("ATTENTION! sacremoses and subword_nmt needs to be installed!")
+            log_line(log)
+            pass
+
+        self.name = model
+        self.layers: List[int] = [int(layer) for layer in layers.split(",")]
+        self.pooling_operation = pooling_operation
+        self.static_embeddings = True
+
+        dummy_sentence: Sentence = Sentence()
+        dummy_sentence.add_token(Token("hello"))
+        embedded_dummy = self.embed(dummy_sentence)
+        self.__embedding_length: int = len(
+            embedded_dummy[0].get_token(1).get_embedding()
+        )
+
+    @property
+    def embedding_length(self) -> int:
+        return self.__embedding_length
+
+    def __getstate__(self):
+        # Copy the object's state from self.__dict__ which contains
+        # all our instance attributes. Always use the dict.copy()
+        # method to avoid modifying the original state.
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        state["model"] = None
+        state["_modules"] = None
+
+        return state
+
+    def __setstate__(self, d):
+        self.__dict__ = d
+        # Restore unpickable entries
+        super().__init__()
+        self.model = torch.hub.load("pytorch/fairseq", self.name)
+
+    def _add_embeddings_internal(self, sentences: List[Sentence]) -> List[Sentence]:
+        self.model.to(flair.device)
+        self.model.eval()
+
+        with torch.no_grad():
+            for sentence in sentences:
+                for token in sentence.tokens:
+                    token_text = token.text
+                    subwords = self.model.encode(token_text)
+
+                    hidden_states = self.model.extract_features(
+                        subwords, return_all_hiddens=True
+                    )
```

Comment:
One other issue is that extracting features per token will probably not produce good representations. Ideally one would encode the entire sentence and realign the representations, similar to what's described here: https://explosion.ai/blog/spacy-pytorch-transformers#alignment.

I will be merging some code very soon that will make this very easy with RoBERTa -- you can provide your own tokenized version of the sentence and it will produced aligned features automatically.

Reply:
@myleott Thanks :) I did a complete re-write of all embedding layers over the last two days. The new strategy is to pass the complete sentence into all models and re-align the embeddings later :)

I'll add some exciting results for NER on CoNLL very soon!